src/violation.rs

```diff
@@ -1,15 +1,98 @@
-// TODO: Implement Violations for constraints
-#[derive(Debug, Clone)]
-pub struct Violations {
-    violations: Vec<Violation>,
+/// Represents Violations found during validation of values based on schema types
+pub trait Violations {
+    /// adds a violation into the collection of [Violation]s
+    fn add_violation(&mut self, violation: ViolationImpl);
+
+    /// Returns `true` if no violations were found, otherwise `false`
+    fn is_valid(&self) -> bool;
+
+    /// Returns the [Violation]s that were found during validation
+    fn violations(&self) -> &[ViolationImpl];
+
+    /// Creates a checkpoint to track the number of violations from this point onwards
+    fn checkpoint(&self) -> Checkpoint;
+}
+
+// Represents collection of [Violation]s
+pub struct ViolationsImpl {
+    violations: Vec<ViolationImpl>,
+}
+
+impl ViolationsImpl {
+    pub fn new() -> Self {
+        Self { violations: vec![] }
+    }
 }
 
-impl Violations {
-    fn add_violation(violation: Violation) {
-        todo!()
+impl Violations for ViolationsImpl {
+    fn add_violation(&mut self, violation: ViolationImpl) {
+        self.violations.push(violation);
     }
+
+    fn is_valid(&self) -> bool {
+        self.violations.is_empty()
+    }
+
+    fn violations(&self) -> &[ViolationImpl] {
+        &self.violations
+    }
+
+    fn checkpoint(&self) -> Checkpoint {
+        Checkpoint::new(self.violations.len())
+    }
+}
+
+// TODO: this can be extended to store children violation count as well for container constraints
+// Represents a checkpoint to be used while validation
+// This determines if children or nested types/constraints have the same number of violations since the checkpoint was created
+pub struct Checkpoint {
+    violations_count: usize,
 }
 
-// TODO: Fill the struct
+impl Checkpoint {
+    pub fn new(violations_count: usize) -> Self {
+        Self { violations_count }
+    }
+
+    pub fn is_valid(&self, violation: &ViolationImpl) -> bool {
+        self.violations_count == violation.violations.len()
+    }
+}
+
+// Represents a single Violation with detailed error message, error code and the constraint for which the validation failed
 #[derive(Debug, Clone)]
-pub struct Violation {}
+pub struct ViolationImpl {
```

Comment:
To me, this name suggests that there is a trait called `Violation`. I'd stick with `pub struct Violation`.

Reply:
Renamed to `ViolationTree` and `VioaltionLeaf`.